File: electron/main/rules/health.ts
import type { RuleAsset, ToolId, ToolSummary } from './types'

export type HealthIssueType = 'oversized' | 'drift' | 'duplicate' | 'missing-entry'

export interface HealthIssue {
  id: string
  type: HealthIssueType
  severity: 'warn' | 'info'
  title: string
  message: string
  assetIds: string[]
  slug?: string
}

export interface HealthReport {
  scannedAt: number
  issueCount: number
  byType: Record<HealthIssueType, number>
  issues: HealthIssue[]
}

const OVERSIZE_BYTES = 30 * 1024

function emptyByType(): Record<HealthIssueType, number> {
  return { oversized: 0, drift: 0, duplicate: 0, 'missing-entry': 0 }
}

/** 基于扫描结果的轻量健康度分析（不二次读盘） */
export function analyzeHealth(tools: ToolSummary[], assets: RuleAsset[]): HealthReport {
  const issues: HealthIssue[] = []

  for (const asset of assets) {
    if (asset.size > OVERSIZE_BYTES) {
      issues.push({
        id: `oversized:${asset.id}`,
        type: 'oversized',
        severity: 'info',
        title: `过大：${asset.title}`,
        message: `${Math.round(asset.size / 1024)} KB，建议拆分或精简（阈值 30KB）`,
        assetIds: [asset.id],
        slug: asset.slug,
      })
    }
  }

  const bySlug = new Map<string, RuleAsset[]>()
  for (const asset of assets) {
    const list = bySlug.get(asset.slug) || []
    list.push(asset)
    bySlug.set(asset.slug, list)
  }

  for (const [slug, list] of bySlug) {
    const toolIds = [...new Set(list.map((a) => a.toolId))]
    if (toolIds.length < 2) continue
    const hashes = list.map((a) => a.contentHash).filter(Boolean) as string[]
    if (hashes.length < 2) continue
    const unique = new Set(hashes)
    if (unique.size > 1) {
      issues.push({
        id: `drift:${slug}`,
        type: 'drift',
        severity: 'warn',
        title: `内容漂移：${list[0]?.title || slug}`,
        message: `在 ${toolIds.join(' / ')} 中同名但内容不一致`,
        assetIds: list.map((a) => a.id),
        slug,
      })
    }
  }

  const byHash = new Map<string, RuleAsset[]>()
  for (const asset of assets) {
    if (!asset.contentHash) continue
    const list = byHash.get(asset.contentHash) || []
    list.push(asset)
    byHash.set(asset.contentHash, list)
  }
  for (const [hash, list] of byHash) {
    if (list.length < 2) continue
    const paths = new Set(list.map((a) => a.absPath))
    if (paths.size < 2) continue
    issues.push({
      id: `dup:${hash}`,
      type: 'duplicate',
      severity: 'info',
      title: `疑似重复：${list[0]?.title || hash}`,
      message: `${list.length} 处内容哈希相同`,
      assetIds: list.map((a) => a.id),
      slug: list[0]?.slug,
    })
  }

  const ENTRY_HINTS: Partial<Record<ToolId, { names: string[]; label: string }>> = {
    claude: { names: ['claude.md', 'agents.md'], label: 'CLAUDE.md / AGENTS.md' },
    cursor: { names: ['agents.md'], label: 'AGENTS.md（可选）' },
    codex: { names: ['agents.md'], label: 'AGENTS.md（可选）' },
  }

  for (const tool of tools) {
    if (!tool.detected) continue
    const hint = ENTRY_HINTS[tool.toolId]
    if (!hint) continue
    const toolAssets = assets.filter((a) => a.toolId === tool.toolId)
    const hasEntry = toolAssets.some((a) => {
      const base = a.relPath.replace(/\\/g, '/').split('/').pop()?.toLowerCase() || ''
      return hint.names.includes(base)
    })
    if (!hasEntry && tool.toolId === 'claude') {
      issues.push({
        id: `missing:${tool.toolId}`,
        type: 'missing-entry',
        severity: 'info',
        title: `缺少入口：${tool.label}`,
        message: `已检测到目录，但未找到 ${hint.label}`,
        assetIds: [],
      })
    }
  }

  const byType = emptyByType()
  for (const issue of issues) byType[issue.type] += 1

  issues.sort((a, b) => {
    if (a.severity !== b.severity) return a.severity === 'warn' ? -1 : 1
    return a.title.localeCompare(b.title)
  })

  return {
    scannedAt: Date.now(),
    issueCount: issues.length,
    byType,
    issues,
  }
}

File: src-tauri/src/rules/health.rs
use std::collections::{HashMap, HashSet};

use super::types::{now_ms, HealthIssue, HealthReport, RuleAsset, ToolSummary};

const OVERSIZE_BYTES: u64 = 30 * 1024;

fn empty_by_type() -> HashMap<String, u32> {
  let mut m = HashMap::new();
  for k in ["oversized", "drift", "duplicate", "missing-entry"] {
    m.insert(k.to_string(), 0);
  }
  m
}

pub fn analyze_health(tools: Vec<ToolSummary>, assets: Vec<RuleAsset>) -> HealthReport {
  let mut issues: Vec<HealthIssue> = Vec::new();

  for asset in &assets {
    if asset.size > OVERSIZE_BYTES {
      issues.push(HealthIssue {
        id: format!("oversized:{}", asset.id),
        issue_type: "oversized".into(),
        severity: "info".into(),
        title: format!("过大：{}", asset.title),
        message: format!(
          "{} KB，建议拆分或精简（阈值 30KB）",
          (asset.size as f64 / 1024.0).round()
        ),
        asset_ids: vec![asset.id.clone()],
        slug: Some(asset.slug.clone()),
      });
    }
  }

  let mut by_slug: HashMap<String, Vec<&RuleAsset>> = HashMap::new();
  for asset in &assets {
    by_slug.entry(asset.slug.clone()).or_default().push(asset);
  }

  for (slug, list) in &by_slug {
    let tool_ids: HashSet<_> = list.iter().map(|a| a.tool_id.as_str()).collect();
    if tool_ids.len() < 2 {
      continue;
    }
    let hashes: Vec<&str> = list
      .iter()
      .filter_map(|a| a.content_hash.as_deref())
      .collect();
    if hashes.len() < 2 {
      continue;
    }
    let unique: HashSet<_> = hashes.iter().copied().collect();
    if unique.len() > 1 {
      let tool_list: Vec<String> = tool_ids.into_iter().map(String::from).collect();
      issues.push(HealthIssue {
        id: format!("drift:{slug}"),
        issue_type: "drift".into(),
        severity: "warn".into(),
        title: format!("内容漂移：{}", list.first().map(|a| a.title.as_str()).unwrap_or(slug)),
        message: format!("在 {} 中同名但内容不一致", tool_list.join(" / ")),
        asset_ids: list.iter().map(|a| a.id.clone()).collect(),
        slug: Some(slug.clone()),
      });
    }
  }

  let mut by_hash: HashMap<String, Vec<&RuleAsset>> = HashMap::new();
  for asset in &assets {
    if let Some(h) = &asset.content_hash {
      by_hash.entry(h.clone()).or_default().push(asset);
    }
  }
  for (hash, list) in &by_hash {
    if list.len() < 2 {
      continue;
    }
    let paths: HashSet<_> = list.iter().map(|a| a.abs_path.as_str()).collect();
    if paths.len() < 2 {
      continue;
    }
    issues.push(HealthIssue {
      id: format!("dup:{hash}"),
      issue_type: "duplicate".into(),
      severity: "info".into(),
      title: format!(
        "疑似重复：{}",
        list.first().map(|a| a.title.as_str()).unwrap_or(hash)
      ),
      message: format!("{} 处内容哈希相同", list.len()),
      asset_ids: list.iter().map(|a| a.id.clone()).collect(),
      slug: list.first().map(|a| a.slug.clone()),
    });
  }

  let entry_hints: HashMap<&str, (&[&str], &str)> = [
    ("claude", (&["claude.md", "agents.md"][..], "CLAUDE.md / AGENTS.md")),
    ("cursor", (&["agents.md"][..], "AGENTS.md（可选）")),
    ("codex", (&["agents.md"][..], "AGENTS.md（可选）")),
  ]
  .into_iter()
  .collect();

  for tool in &tools {
    if !tool.detected {
      continue;
    }
    let Some((names, label)) = entry_hints.get(tool.tool_id.as_str()) else {
      continue;
    };
    let tool_assets: Vec<_> = assets.iter().filter(|a| a.tool_id == tool.tool_id).collect();
    let has_entry = tool_assets.iter().any(|a| {
      let base = a
        .rel_path
        .replace('\\', "/")
        .split('/')
        .next_back()
        .unwrap_or("")
        .to_lowercase();
      names.iter().any(|n| *n == base)
    });
    if !has_entry && tool.tool_id == "claude" {
      issues.push(HealthIssue {
        id: format!("missing:{}", tool.tool_id),
        issue_type: "missing-entry".into(),
        severity: "info".into(),
        title: format!("缺少入口：{}", tool.label),
        message: format!("已检测到目录，但未找到 {label}"),
        asset_ids: vec![],
        slug: None,
      });
    }
  }

  let mut by_type = empty_by_type();
  for issue in &issues {
    *by_type.entry(issue.issue_type.clone()).or_insert(0) += 1;
  }

  issues.sort_by(|a, b| {
    if a.severity != b.severity {
      return if a.severity == "warn" {
        std::cmp::Ordering::Less
      } else {
        std::cmp::Ordering::Greater
      };
    }
    a.title.cmp(&b.title)
  });

  HealthReport {
    scanned_at: now_ms(),
    issue_count: issues.len() as u32,
    by_type,
    issues,
  }
}
